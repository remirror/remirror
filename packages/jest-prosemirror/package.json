{
  "name": "jest-prosemirror",
  "version": "3.0.0-beta.4",
  "description": "Write expressive tests for your prosemirror editor",
  "keywords": [
    "jest",
    "prosemirror",
    "snapshots",
    "tests"
  ],
  "homepage": "https://github.com/remirror/remirror/tree/HEAD/packages/jest-prosemirror",
  "repository": {
    "type": "git",
    "url": "https://github.com/remirror/remirror.git",
    "directory": "packages/jest-prosemirror"
  },
  "license": "MIT",
  "contributors": [
    "Ifiok Jr. <ifiokotung@gmail.com>"
  ],
  "type": "module",
  "exports": {
    ".": {
      "types": "./dist-types/index.d.ts",
      "import": "./dist/jest-prosemirror.js",
      "require": "./dist/jest-prosemirror.cjs"
    },
    "./environment": {
      "types": "./dist-types/environment.d.ts",
      "import": "./environment/dist/jest-prosemirror-environment.js",
      "require": "./environment/dist/jest-prosemirror-environment.cjs"
    },
    "./package.json": "./package.json",
    "./serializer": {
      "types": "./dist-types/serializer.d.ts",
      "require": "./serializer/dist/jest-prosemirror-serializer.cjs"
    }
  },
  "main": "./dist/jest-prosemirror.cjs",
  "module": "./dist/jest-prosemirror.js",
  "types": "./dist-types/index.d.ts",
  "files": [
    "dist",
    "dist-types",
    "environment",
    "serializer"
  ],
  "scripts": {
    "build": "remirror-cli build"
  },
  "dependencies": {
    "@babel/runtime": "^7.22.3",
    "@remirror/core-constants": "3.0.0-beta.3",
    "@remirror/core-helpers": "4.0.0-beta.3",
    "@remirror/core-types": "3.0.0-beta.4",
    "@remirror/core-utils": "3.0.0-beta.4",
    "@remirror/pm": "3.0.0-beta.4",
    "@testing-library/dom": "^9.3.0",
    "@types/stringify-object": "^3.3.1",
    "chalk": "^4.1.0",
    "jest-matcher-utils": "^29.5.0",
    "prosemirror-schema-basic": "^1.2.3",
    "prosemirror-tables": "^1.3.7",
    "prosemirror-test-builder": "^1.1.1",
    "stringify-object": "^3.3.0",
    "test-keyboard": "2.0.7-beta.4"
  },
  "devDependencies": {
    "@remirror/cli": "1.1.0-beta.0",
    "jest": "^29.5.0"
  },
  "peerDependencies": {
    "jest": ">= 24.9.0"
  },
  "peerDependenciesMeta": {
    "jest": {
      "optional": true
    }
  },
  "publishConfig": {
    "access": "public",
    "exports": {
      ".": {
        "types": "./dist/jest-prosemirror.d.ts",
        "import": "./dist/jest-prosemirror.js",
        "require": "./dist/jest-prosemirror.cjs"
      },
      "./environment": {
        "types": "./environment/dist/jest-prosemirror-environment.d.ts",
        "import": "./environment/dist/jest-prosemirror-environment.js",
        "require": "./environment/dist/jest-prosemirror-environment.cjs"
      },
      "./package.json": "./package.json",
      "./serializer": {
        "types": "./serializer/dist/jest-prosemirror-serializer.d.ts",
        "require": "./serializer/dist/jest-prosemirror-serializer.cjs"
      }
    },
    "main": "./dist/jest-prosemirror.cjs",
    "module": "./dist/jest-prosemirror.js",
    "type": "module",
    "types": "./dist/jest-prosemirror.d.ts"
  },
  "@remirror": {
    "tsconfigs": {
      "src": {
        "compilerOptions": {
          "types": [
            "node",
            "jest"
          ]
        }
      }
    }
  },
  "preconstruct": {
    "entrypoints": [
      "index.ts",
      "environment.ts",
      "serializer.cts"
    ]
  }
}
