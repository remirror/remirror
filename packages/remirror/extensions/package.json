{
  "type": "module",
  "exports": {
    ".": {
      "types": "./dist/remirror-extensions.d.ts",
      "import": "./dist/remirror-extensions.js",
      "require": "./dist/remirror-extensions.cjs"
    },
    "./package.json": "./package.json"
  },
  "main": "./dist/remirror-extensions.cjs",
  "module": "./dist/remirror-extensions.js",
  "types": "./dist/remirror-extensions.d.ts"
}
