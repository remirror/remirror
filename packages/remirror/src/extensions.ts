// Extensions
export * from '@remirror/extension-annotation';
export * from '@remirror/extension-bidi';
export * from '@remirror/extension-blockquote';
export * from '@remirror/extension-bold';
export * from '@remirror/extension-callout';
export * from '@remirror/extension-code';
export * from '@remirror/extension-code-block';
export * from '@remirror/extension-codemirror5';
export * from '@remirror/extension-collaboration';
export * from '@remirror/extension-columns';
export * from '@remirror/extension-diff';
export * from '@remirror/extension-doc';
export * from '@remirror/extension-drop-cursor';
export * from '@remirror/extension-embed';
export * from '@remirror/extension-emoji';
export * from '@remirror/extension-entity-reference';
export * from '@remirror/extension-epic-mode';
export * from '@remirror/extension-events';
export * from '@remirror/extension-find';
export * from '@remirror/extension-font-family';
export * from '@remirror/extension-font-size';
export * from '@remirror/extension-gap-cursor';
export * from '@remirror/extension-hard-break';
export * from '@remirror/extension-heading';
export * from '@remirror/extension-history';
export * from '@remirror/extension-horizontal-rule';
export * from '@remirror/extension-image';
export * from '@remirror/extension-italic';
export * from '@remirror/extension-link';
export * from '@remirror/extension-list';
export * from '@remirror/extension-markdown';
export * from '@remirror/extension-mention';
export * from '@remirror/extension-mention-atom';
export * from '@remirror/extension-node-formatting';
export * from '@remirror/extension-paragraph';
export * from '@remirror/extension-placeholder';
export * from '@remirror/extension-positioner';
export * from '@remirror/extension-shortcuts';
export * from '@remirror/extension-strike';
export * from '@remirror/extension-sub';
export * from '@remirror/extension-sup';
export * from '@remirror/extension-tables';
export * from '@remirror/extension-text';
export * from '@remirror/extension-text-case';
export * from '@remirror/extension-text-color';
export * from '@remirror/extension-text-highlight';
export * from '@remirror/extension-trailing-node';
export * from '@remirror/extension-underline';
export * from '@remirror/extension-whitespace';
export * from '@remirror/extension-yjs';

// Presets
export * from '@remirror/preset-core';
export * from '@remirror/preset-formatting';
export * from '@remirror/preset-wysiwyg';
