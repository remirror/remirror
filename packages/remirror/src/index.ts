export * from '@remirror/core';
export * from '@remirror/theme';
