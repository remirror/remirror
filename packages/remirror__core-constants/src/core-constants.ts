/**
 * The css class added to a node that is selected.
 */
export const SELECTED_NODE_CLASS_NAME = 'ProseMirror-selectednode';

/**
 * The css selector for a selected node.
 */
export const SELECTED_NODE_CLASS_SELECTOR = `.${SELECTED_NODE_CLASS_NAME}`;

/**
 * ProseMirror uses the Unicode Character 'OBJECT REPLACEMENT CHARACTER'
 * (U+FFFC) as text representation for leaf nodes, i.e. nodes that don't have
 * any content or text property (e.g. hardBreak, emoji, mention, rule) It was
 * introduced because of https://github.com/ProseMirror/prosemirror/issues/262
 * This can be used in an input rule regex to be able to include or exclude such
 * nodes.
 */
export const LEAF_NODE_REPLACING_CHARACTER = '\uFFFC';

/**
 * The null character.
 *
 * See {@link https://stackoverflow.com/a/6380172}
 */
export const NULL_CHARACTER = '\0';

/**
 * Indicates that a state update was caused by an override and not via
 * transactions or user commands.
 *
 * This is the case when `setContent` is called and for all `controlled` updates
 * within a `react` editor instance.
 */
export const STATE_OVERRIDE = '__state_override__';

/**
 * The global name for the module exported by the remirror webview bundle.
 */
export const REMIRROR_WEBVIEW_NAME = '$$__REMIRROR_WEBVIEW_BUNDLE__$$';

/**
 * A character useful for separating inline nodes.
 *
 * @remarks
 * Typically used in decorations as follows.
 *
 * ```ts
 * document.createTextNode(ZERO_WIDTH_SPACE_CHAR);
 * ```
 *
 * This produces the html entity '8203'
 */
export const ZERO_WIDTH_SPACE_CHAR = '\u200B';

/**
 * The non breaking space character.
 */
export const NON_BREAKING_SPACE_CHAR = '\u00A0';

/**
 * A default empty object node. Useful for resetting the content of a
 * prosemirror document.
 */
export const EMPTY_PARAGRAPH_NODE = {
  type: 'doc',
  content: [
    {
      type: 'paragraph',
    },
  ],
};

export const EMPTY_NODE = {
  type: 'doc',
  content: [],
};

/**
 * The type for the extension tags..
 */
export type ExtensionTag = Remirror.ExtensionTags & typeof BaseExtensionTag;

/**
 * A method for updating the extension tags.
 *
 * ```tsx
 * import { ExtensionTag, mutateTag } from 'remirror';
 *
 * mutateTag((tag) => {
 *   tag.SuperCustom = 'superCustom';
 * });
 *
 * declare global {
 *   namespace Remirror {
 *     interface ExtensionTag {
 *       SuperCustom: 'superCustom';
 *     }
 *   }
 * }
 *
 *
 * log(ExtensionTag.SuperCustom); // This is fine ✅
 * log(ExtensionTag.NotDefined); // This will throw ❌
 * ```
 */
export function mutateTag(mutator: (Tag: ExtensionTag) => void): void {
  mutator(BaseExtensionTag);
}

const BaseExtensionTag = {
  /**
   * Describes a node that can be used as the last node of a document and
   * doesn't need to have anything else rendered after itself.
   *
   * @remarks
   *
   * e.g. `paragraph`
   */
  LastNodeCompatible: 'lastNodeCompatible',

  /**
   * A mark that is used to change the formatting of the node it wraps.
   *
   * @remarks
   *
   * e.g. `bold`, `italic`
   */
  FormattingMark: 'formattingMark',

  /**
   * A node that formats text in a non-standard way.
   *
   * @remarks
   *
   * e.g. `codeBlock`, `heading`, `blockquote`
   */
  FormattingNode: 'formattingNode',

  /**
   * Identifies a node which has problems with cursor navigation.
   *
   * @remarks
   *
   * When this tag is added to an extension this will be picked up by
   * behavioural extensions such as the NodeCursorExtension which makes hard to
   * reach nodes reachable using keyboard arrows.
   */
  NodeCursor: 'nodeCursor',

  /**
   * Mark group for font styling (e.g. bold, italic, underline, superscript).
   */
  FontStyle: 'fontStyle',

  /**
   * Mark groups for links.
   */
  Link: 'link',

  /**
   * Mark groups for colors (text-color, background-color, etc).
   */
  Color: 'color',

  /**
   * Mark group for alignment.
   */
  Alignment: 'alignment',

  /**
   * Mark group for indentation.
   */
  Indentation: 'indentation',

  /**
   * Extension which affect the behaviour of the content. Can be nodes marks or
   * plain.
   */
  Behavior: 'behavior',

  /**
   * Marks and nodes which contain code.
   */
  Code: 'code',

  /**
   * Whether this node is an inline node.
   *
   * - `text` is an inline node, but `paragraph` is a block node.
   */
  InlineNode: 'inline',

  /**
   * This is a node that can contain list items.
   */
  ListContainerNode: 'listContainer',

  /**
   * Tags the extension as a list item node which can be contained by
   * [[`ExtensionTag.ListNode`]].
   */
  ListItemNode: 'listItemNode',

  /**
   * Sets this as a block level node.
   */
  Block: 'block',

  /**
   * @deprecate use `ExtensionTags.Block` instead.
   */
  BlockNode: 'block',

  /**
   * Set this as a text block
   */
  TextBlock: 'textBlock',

  /**
   * A tag that excludes this from input rules.
   */
  ExcludeInputRules: 'excludeFromInputRules',

  /**
   * A mark or node that can't  be exited when at the end and beginning of the
   * document with an arrow key or backspace key.
   */
  PreventExits: 'preventsExits',

  /**
   * Represents a media compatible node.
   */
  Media: 'media',
} as const;

/**
 * These are the default supported tag strings which help categorize different
 * behaviors that extensions can exhibit.
 *
 * @remarks
 *
 * Any extension can register itself with multiple such behaviors and these
 * categorizations can be used by other extensions when running commands and
 * updating the document.
 */
export const ExtensionTag = BaseExtensionTag as ExtensionTag;

/**
 * The string values which can be used as extension tags.
 */
export type ExtensionTagType = ExtensionTag[keyof ExtensionTag];

/**
 * The identifier key which is used to check objects for whether they are a
 * certain type.
 *
 * @remarks
 *
 * Just pretend you don't know this exists.
 *
 * @internal
 */
export const __INTERNAL_REMIRROR_IDENTIFIER_KEY__: unique symbol = Symbol.for('__remirror__');

/**
 * These constants are stored on the `REMIRROR_IDENTIFIER_KEY` property of
 * `remirror` related constructors and instances in order to identify them as
 * being internal to Remirror.
 *
 * @remarks
 *
 * This helps to prevent issues around check types via `instanceof` which can
 * lead to false negatives.
 *
 * @internal
 */
export enum RemirrorIdentifier {
  /**
   * Identifies `PlainExtension`s.
   */
  PlainExtension = 'RemirrorPlainExtension',

  /**
   * Identifies `NodeExtension`s.
   */
  NodeExtension = 'RemirrorNodeExtension',

  /**
   * Identifies `MarkExtension`s.
   */
  MarkExtension = 'RemirrorMarkExtension',

  /**
   * Identifies `PlainExtensionConstructor`s.
   */
  PlainExtensionConstructor = 'RemirrorPlainExtensionConstructor',

  /**
   * Identifies `NodeExtensionConstructor`s.
   */
  NodeExtensionConstructor = 'RemirrorNodeExtensionConstructor',

  /**
   * Identifies `MarkExtensionConstructor`s.
   */
  MarkExtensionConstructor = 'RemirrorMarkExtensionConstructor',

  /**
   * The string used to identify an instance of the `Manager`
   */
  Manager = 'RemirrorManager',

  /**
   * The preset type identifier.
   */
  Preset = 'RemirrorPreset',

  /**
   * The preset type identifier.
   */
  PresetConstructor = 'RemirrorPresetConstructor',
}

/**
 * The priority of extension which determines what order it is loaded into the
 * editor.
 *
 * @remarks
 *
 * Higher priority extension (higher numberic value) will ensure the extension
 * has a higher preference in your editor. In the case where you load two
 * identical extensions into your editor (same name, or same constructor), the
 * extension with the  higher priority is the one that will be loaded.
 *
 * The higher the numeric value the higher the priority. The priority can also
 * be passed a number but naming things in this `enum` should help provide some
 * context to the numbers.
 *
 * By default all extensions are created with a `ExtensionPriority.Default`.
 */
export enum ExtensionPriority {
  /**
   * Use this **never** 😉
   */
  Critical = 1_000_000,

  /**
   * A, like super duper, high priority.
   */
  Highest = 100_000,

  /**
   * The highest priority level that should be used in a publicly shared
   * extension (to allow some wiggle room for downstream users overriding
   * priorities).
   */
  High = 10_000,

  /**
   * A medium priority extension. This is typically all you need to take
   * priority over built in extensions.
   */
  Medium = 1000,

  /**
   * This is the **default** priority for most extensions.
   */
  Default = 100,

  /**
   * This is the **default** priority for builtin behavior changing extensions.
   */
  Low = 10,

  /**
   * This is useful for extensions that exist to be overridden.
   */
  Lowest = 0,
}

/**
 * Identifies the stage the extension manager is at.
 */
export enum ManagerPhase {
  /**
   * The initial value for the manager phase.
   */
  None,

  /**
   * When the extension manager is being created and the onCreate methods are
   * being called.
   *
   * This happens within the RemirrorManager constructor.
   */
  Create,

  /**
   * When the view is being added and all `onView` lifecycle methods are being
   * called. The view is typically added before the dom is ready for it.
   */
  EditorView,

  /**
   * The phases of creating this manager are completed and `onTransaction` is
   * called every time the state updates.
   */
  Runtime,

  /**
   * The manager is being destroyed.
   */
  Destroy,
}

/**
 * The named shortcuts that can be used to update multiple commands.
 */
export enum NamedShortcut {
  Undo = '_|undo|_',
  Redo = '_|redo|_',
  Bold = '_|bold|_',
  Italic = '_|italic|_',
  Underline = '_|underline|_',
  Strike = '_|strike|_',
  Code = '_|code|_',
  Paragraph = '_|paragraph|_',
  H1 = '_|h1|_',
  H2 = '_|h2|_',
  H3 = '_|h3|_',
  H4 = '_|h4|_',
  H5 = '_|h5|_',
  H6 = '_|h6|_',
  TaskList = '_|task|_',
  BulletList = '_|bullet|_',
  OrderedList = '_|number|_',
  Quote = '_|quote|_',
  Divider = '_|divider|_',
  Codeblock = '_|codeblock|_',
  ClearFormatting = '_|clear|_',
  Superscript = '_|sup|_',
  Subscript = '_|sub|_',
  LeftAlignment = '_|left-align|_',
  CenterAlignment = '_|center-align|_',
  RightAlignment = '_|right-align|_',
  JustifyAlignment = '_|justify-align|_',
  InsertLink = '_|link|_',
  /** @deprecated */
  Find = '_|find|_',
  /** @deprecated */
  FindBackwards = '_|find-backwards|_',
  /** @deprecated */
  FindReplace = '_|find-replace|_',
  AddFootnote = '_|footnote|_',
  AddComment = '_|comment|_',
  ContextMenu = '_|context-menu|_',
  IncreaseFontSize = '_|inc-font-size|_',
  DecreaseFontSize = '_|dec-font-size|_',
  IncreaseIndent = '_|indent|_',
  DecreaseIndent = '_|dedent|_',
  Shortcuts = '_|shortcuts|_',
  Copy = '_|copy|_',
  Cut = '_|cut|_',
  Paste = '_|paste|_',
  PastePlain = '_|paste-plain|_',
  SelectAll = '_|select-all|_',

  /**
   * A keyboard shortcut to trigger formatting the current block.
   *
   * @defaultValue 'Alt-Shift-F' (Mac) | 'Shift-Ctrl-F' (PC)
   */
  Format = '_|format|_',
}

/**
 * Helpful empty array for use when a default array value is needed.
 *
 * DO NOT MUTATE!
 */
export const EMPTY_ARRAY: never[] = [];

declare global {
  namespace Remirror {
    /**
     * This interface is for extending the default `ExtensionTag`'s in your
     * codebase with full type checking support.
     */
    interface ExtensionTags {}
  }
}
