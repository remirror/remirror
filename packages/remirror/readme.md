# remirror

> One package to rule them all, one entry point to bind them.

[![Version][version]][npm] [![Weekly Downloads][downloads-badge]][npm] [![Bundled size][size-badge]][size] [![Typed Codebase][typescript]](#) [![MIT License][license]](#)

[version]: https://flat.badgen.net/npm/v/remirror
[npm]: https://npmjs.com/package/remirror
[license]: https://flat.badgen.net/badge/license/MIT/purple
[size]: https://bundlephobia.com/result?p=remirror
[size-badge]: https://flat.badgen.net/bundlephobia/minzip/remirror
[typescript]: https://flat.badgen.net/badge/icon/TypeScript?icon=typescript&label
[downloads-badge]: https://badgen.net/npm/dw/remirror/red?icon=npm

## Installation

```bash
# yarn
yarn add remirror @remirror/pm

# pnpm
pnpm add remirror @remirror/pm

# npm
npm install remirror @remirror/pm
```

The `remirror` package will automatically install the `@remirror/pm` package for you. You only need to install it yourself if you'd like to ensure consistent versions of the `prosemirror-*` libraries when importing from `@remirror/pm/state` instead of `prosemirror-state` or `@remirror/pm/model` instead of `prosemirror-model`.

## Usage

Rather than installing multiple scoped packages, the `remirror` package is a gateway to using all the goodness that remirror provides while minimising your bundle size.

The following creates a controlled editor with React.

```tsx
import React from 'react';
import { SocialPreset } from 'remirror/extensions';
import { Remirror, SocialEmojiComponent, useRemirror } from '@remirror/react';

const EditorWrapper = () => {
  const socialPreset = new SocialPreset();
  const { state, onChange } = useRemirror({ extensions: () => [...socialPreset()] });

  return (
    <Remirror state={state} onChange={onChange} manager={manager} autoRender={true}>
      <SocialEmojiComponent />
    </Remirror>
  );
};
```

These are the entry points available through the `remirror` package.

- `remirror` - All the core functionality available through `@remirror/core`.
- `remirror/extensions` - All the core extensions and presets made available through the main `remirror` repository. This doesn't include any framework specific extensions and presets.
- `remirror/dom` - The dom framework implementation of via `createDomEditor`.
