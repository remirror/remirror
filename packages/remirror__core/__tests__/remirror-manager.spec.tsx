import { jest } from '@jest/globals';
import { createEditor, doc, p } from 'jest-prosemirror';
import {
  isRemirrorManager,
  NodeExtension,
  NodeViewsExtension,
  PlainExtension,
  RemirrorManager,
} from 'remirror';
import { corePreset, createCoreManager, HeadingExtension } from 'remirror/extensions';
import { createFramework, hideConsoleError } from 'testing';
import { EMPTY_PARAGRAPH_NODE, ExtensionPriority, ExtensionTag } from '@remirror/core-constants';
import type {
  Dispose,
  KeyBindingCommandFunction,
  NodeExtensionSpec,
  NodeViewMethod,
  ProsemirrorAttributes,
} from '@remirror/core-types';
import { Schema } from '@remirror/pm/model';
import { EditorState, Plugin, Transaction } from '@remirror/pm/state';
import { EditorView } from '@remirror/pm/view';

describe('Manager', () => {
  let state: EditorState;

  const innerMock: any = jest.fn();
  const mock: any = jest.fn((_: ProsemirrorAttributes) => innerMock);
  const getInformation: any = jest.fn(() => 'information');

  class DummyExtension extends PlainExtension {
    get name() {
      return 'dummy' as const;
    }

    createTags() {
      return [ExtensionTag.Behavior, ExtensionTag.LastNodeCompatible];
    }

    createCommands() {
      return { dummy: mock };
    }

    createHelpers() {
      return {
        getInformation,
      };
    }

    createAttributes() {
      return {
        class: 'custom',
      };
    }
  }

  class BigExtension extends NodeExtension {
    static disableExtraAttributes = true;

    get name() {
      return 'big' as const;
    }

    createNodeSpec(): NodeExtensionSpec {
      return {
        toDOM: () => ['h1', 0],
      };
    }
  }

  const dummyExtension = new DummyExtension({ priority: ExtensionPriority.Critical });
  const bigExtension = new BigExtension({ priority: ExtensionPriority.Lowest });

  let manager = RemirrorManager.create([dummyExtension, bigExtension, ...corePreset()]);

  let view: EditorView;

  beforeEach(() => {
    manager = RemirrorManager.create(() => [dummyExtension, bigExtension, ...corePreset()]);
    manager.attachFramework(createFramework(manager), () => {});
    state = manager.createState({ content: EMPTY_PARAGRAPH_NODE });
    view = new EditorView(document.createElement('div'), {
      state,
      editable: () => true,
      plugins: [],
    });
    manager.addView(view);
  });

  hideConsoleError(true);

  it('enforces constructor privacy', () => {
    // @ts-expect-error
    expect(() => new RemirrorManager({})).toThrow();
  });

  it('supports commands', () => {
    const attributes = { a: 'a' };
    manager.store.commands.dummy(attributes);
    const { state, dispatch } = manager.view;

    expect(mock).toHaveBeenCalledWith(attributes);
    expect(innerMock).toHaveBeenCalledWith({
      state: state,
      dispatch: dispatch,
      view: manager.view,
      tr: expect.any(Transaction),
    });

    const { tr } = innerMock.mock.calls[0][0];
    // Check this transaction is empty
    expect(tr.steps).toHaveLength(0);
  });

  it('supports helpers', () => {
    const value = manager.store.helpers.getInformation();

    expect(value).toBe('information');
    expect(getInformation).toHaveBeenCalled();
  });

  describe('#properties', () => {
    it('should sort extensions by priority', () => {
      expect(manager.extensions[0]?.name).toBe('dummy');
      expect(manager.extensions[manager.extensions.length - 1]?.name).toBe('big');
    });

    it('should allow overriding the priority', () => {
      manager = manager.recreate([], { priority: { dummy: ExtensionPriority.Lowest } });
      expect(manager.extensions[0]?.name).not.toBe('dummy');
      expect(manager.extensions[manager.extensions.length - 1]?.name).toBe('big');
      expect(manager.extensions[manager.extensions.length - 2]?.name).toBe('dummy');
    });

    it('should provide the schema at instantiation', () => {
      expect(createCoreManager([]).schema).toBeInstanceOf(Schema);
    });

    it('should provide access to `attributes`', () => {
      expect(manager.store.attributes.class).toInclude('custom');
    });

    hideConsoleError(true);

    it('should initialize the `extensionStore` with the correct values', () => {
      const manager = createCoreManager([]);

      expect(Object.keys(manager.extensionStore)).toMatchSnapshot();
      expect(manager.extensionStore.previousState).toBeUndefined();
      expect(() => manager.extensionStore.currentState).toThrowErrorMatchingSnapshot();
      expect(manager.extensionStore.getState).toEqual(expect.any(Function));
    });
  });

  it('isManager', () => {
    expect(isRemirrorManager({})).toBeFalse();
    expect(isRemirrorManager(null)).toBeFalse();
    expect(isRemirrorManager(dummyExtension)).toBeFalse();
    expect(isRemirrorManager(manager, ['dummy', 'biggest'])).toBeFalse();
    expect(isRemirrorManager(manager, [class A extends DummyExtension {}])).toBeFalse();
    expect(isRemirrorManager(manager)).toBeTrue();
    expect(isRemirrorManager(manager, [DummyExtension])).toBeTrue();
    expect(isRemirrorManager(manager, ['dummy', 'big'])).toBeTrue();
  });

  it('output', () => {
    const manager = createCoreManager([]);
    expect(manager.output).toBeUndefined();
    const framework = createFramework(manager);

    manager.attachFramework(framework, () => {});
    expect(manager.output).toBe(framework.frameworkOutput);
  });
});

test('keymaps', () => {
  const mocks = {
    firstEnter: jest.fn((..._: Parameters<KeyBindingCommandFunction>) => false),
    secondEnter: jest.fn((..._: Parameters<KeyBindingCommandFunction>) => false),
    thirdEnter: jest.fn((..._: Parameters<KeyBindingCommandFunction>) => false),
  };

  class FirstExtension extends PlainExtension {
    get name() {
      return 'first' as const;
    }

    createKeymap() {
      return {
        Enter: mocks.firstEnter,
      };
    }
  }

  class SecondExtension extends PlainExtension {
    get name() {
      return 'second' as const;
    }

    createKeymap() {
      return {
        Enter: mocks.secondEnter,
      };
    }
  }

  class ThirdExtension extends PlainExtension {
    get name() {
      return 'third' as const;
    }

    createKeymap() {
      return {
        Enter: mocks.thirdEnter,
      };
    }
  }

  const manager = RemirrorManager.create(() => [
    new FirstExtension(),
    new SecondExtension(),
    new ThirdExtension(),
    ...corePreset(),
  ]);
  manager.attachFramework(createFramework(manager), () => {});

  createEditor(doc(p('simple<cursor>')), { plugins: manager.store.plugins })
    .insertText('abcd')
    .press('Enter')
    .callback(() => {
      expect(mocks.firstEnter).toHaveBeenCalled();
      expect(mocks.secondEnter).toHaveBeenCalled();
      expect(mocks.thirdEnter).toHaveBeenCalled();

      jest.clearAllMocks();
      mocks.firstEnter.mockImplementation(() => true);
    })
    .press('Enter')
    .callback(() => {
      expect(mocks.firstEnter).toHaveBeenCalled();
      expect(mocks.secondEnter).not.toHaveBeenCalled();
      expect(mocks.thirdEnter).not.toHaveBeenCalled();

      jest.clearAllMocks();
      mocks.firstEnter.mockImplementation(({ next }) => {
        return next();
      });
    })
    .press('Enter')
    .callback(() => {
      expect(mocks.secondEnter).toHaveBeenCalledTimes(1);
      expect(mocks.thirdEnter).toHaveBeenCalledTimes(1);

      jest.clearAllMocks();
      mocks.secondEnter.mockImplementation(() => true);
    })
    .press('Enter')
    .callback(() => {
      expect(mocks.secondEnter).toHaveBeenCalledTimes(1);
      expect(mocks.thirdEnter).not.toHaveBeenCalled();
    });
});

test('lifecycle', () => {
  expect.assertions(5);

  class LifecycleExtension extends PlainExtension {
    static defaultPriority = ExtensionPriority.Lowest;

    get name() {
      return 'test' as const;
    }

    onCreate() {
      expect(this.store.setExtensionStore).toBeFunction();
      expect(this.store.setStoreKey).toBeFunction();
      expect(this.store.getStoreKey).toBeFunction();
      expect(this.store.tags).toBeObject();
      expect(this.store.schema).toBeInstanceOf(Schema);
    }
  }

  const extension = new LifecycleExtension();
  createCoreManager([extension]);
});

describe('createEmptyDoc', () => {
  it('can create an empty doc', () => {
    const manager = RemirrorManager.create([...corePreset()]);

    expect(manager.createEmptyDoc().toJSON()).toMatchInlineSnapshot(`
      {
        "content": [
          {
            "type": "paragraph",
          },
        ],
        "type": "doc",
      }
    `);
  });

  it('creates an empty doc with alternative content', () => {
    const headingManager = RemirrorManager.create([
      ...corePreset({ content: 'heading+' }),
      new HeadingExtension(),
    ]);
    expect(headingManager.createEmptyDoc()).toMatchInlineSnapshot(`
      Prosemirror node: {
        "type": "doc",
        "content": [
          {
            "type": "heading",
            "attrs": {
              "level": 1
            }
          }
        ]
      }
    `);
  });
});

describe('options', () => {
  it('can add extra plugins', () => {
    const extraPlugin = new Plugin({});
    const manager = createCoreManager([], { plugins: [extraPlugin] });

    expect(manager.store.plugins).toContain(extraPlugin);
  });

  it('can add additional nodeViews', () => {
    const custom: NodeViewMethod = jest.fn(() => ({})) as unknown as NodeViewMethod;
    const nodeViews: Record<string, NodeViewMethod> = { custom };
    const manager = createCoreManager([], { nodeViews });

    expect(manager.getExtension(NodeViewsExtension).plugin.spec.props?.nodeViews?.custom).toBe(
      custom,
    );
  });
});

test('disposes of methods', () => {
  const mocks = {
    create: jest.fn(),
    view: jest.fn(),
  };

  class DisposeExtension extends PlainExtension {
    get name() {
      return 'dispose' as const;
    }

    onCreate(): Dispose {
      return mocks.create;
    }

    onView(): Dispose {
      return mocks.view;
    }
  }

  const manager = RemirrorManager.create(() => [new DisposeExtension(), ...corePreset()]);
  const framework = createFramework(manager);

  const view = new EditorView(document.createElement('div'), {
    state: framework.initialEditorState,
    editable: () => true,
    plugins: [],
  });

  manager.attachFramework(framework, () => {});
  manager.addView(view);

  manager.destroy();

  expect(mocks.create).toHaveBeenCalledTimes(1);
  expect(mocks.view).toHaveBeenCalledTimes(1);
});
