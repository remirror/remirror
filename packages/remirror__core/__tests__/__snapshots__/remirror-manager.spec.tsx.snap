// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Manager #properties should initialize the \`extensionStore\` with the correct values 1`] = `
[
  "extensions",
  "phase",
  "view",
  "managerSettings",
  "getState",
  "updateState",
  "isMounted",
  "getExtension",
  "manager",
  "document",
  "stringHandlers",
  "currentState",
  "previousState",
  "getStoreKey",
  "setStoreKey",
  "setExtensionStore",
  "setStringHandler",
  "nodeNames",
  "markNames",
  "plainNames",
  "tags",
  "plainTags",
  "markTags",
  "nodeTags",
  "schema",
  "updatePlugins",
  "dispatchPluginUpdate",
  "updateExtensionPlugins",
  "getPluginState",
  "updateAttributes",
  "rebuildInputRules",
  "addSuggester",
  "removeSuggester",
  "attrs",
  "active",
  "helpers",
  "rebuildKeymap",
  "createPlaceholderCommand",
]
`;

exports[`Manager #properties should initialize the \`extensionStore\` with the correct values 2`] = `
"This occurs when accessing a method or property before it is available.

\`getState\` can only be called after the \`Framework\` or the \`EditorView\` has been added to the manager\`. Check your plugins to make sure that the decorations callback uses the state argument.

For more information visit https://remirror.io/docs/errors#rmr0008"
`;
